Replace deprecated Class.newInstance()

SethSmucker opened the pull request:
## Summary
- Replace deprecated `Class.newInstance()` with `getDeclaredConstructor().newInstance()`
- Update exception handling to use `ReflectiveOperationException` (covers the additional exceptions from the new API)

## Files Changed
- `DatawaveMetricsConfig.java`
- `ProtostuffHttpMessageConverter.java`

Fixes #3311
Part of #2443

SethSmucker committed:
Replace deprecated Class.newInstance() with getDeclaredConstructor().newInstance()

Class.newInstance() has been deprecated since Java 9 because it
propagates exceptions thrown by the nullary constructor, including
checked exceptions. The replacement getDeclaredConstructor().newInstance()
wraps such exceptions in an InvocationTargetException.

foster33 commented:
There are additional occurrences of `Class.newInstance()` throughout the entire code base outside of just `spring-boot-starter-datawave`. Those should probably be added as well

SethSmucker committed:
Replace remaining deprecated Class.newInstance() calls

- QueryIterator.java: c.newInstance() -> c.getDeclaredConstructor().newInstance()
- AggregatingReducer.java: clazz.newInstance() -> clazz.getDeclaredConstructor().newInstance()
- Updated exception handling in related methods and overrides

SethSmucker committed:
Merge branch 'integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

foster33 committed:
Merge branch 'integration' into task/replace-deprecated-class-newinstance

foster33 reviewed (changes requested):
There seems to be 4 more uses of `Class.newInstance()`. I will go ahead and make those changes.

foster33 committed:
update additional uses of deprecated newinstance

apmoriarty commented on warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/job/reduce/AggregatingReducer.java:
```diff
@@ -439,11 +439,12 @@ public CustomColumnToClassMapping(Map<String,Entry<Map<String,String>,String>> c
 
-                } catch (InstantiationException | IllegalAccessException | ClassNotFoundException | IOException e) {
+                } catch (InstantiationException | IllegalAccessException | ClassNotFoundException | IOException | NoSuchMethodException
+                                | java.lang.reflect.InvocationTargetException e) {
```
No need for the fully qualified import. Bad bot.

foster33 committed:
Remove fully qualified imports for InvocationTargetException

lbschanno commented:
One additional usage that needs to be replaced:
- BulkIngestKeyAggregatingReducer (line 66)

SethSmucker committed:
Merge integration into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote changes

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

foster33 committed:
Merge branch 'integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/replace-deprecated-class-newinstance

foster33 commented:
Seems like the usage that Laura mentioned is still there:

`BulkIngestKeyAggregatingReducer` line 66

SethSmucker committed:
Merge branch 'integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Merge branch 'integration' into task/replace-deprecated-class-newinstance

SethSmucker committed:
Replace Class.newInstance() in BulkIngestKeyAggregatingReducer
